packages/language-server/src/format/comments.ts

```diff
@@ -0,0 +1,113 @@
+    if (Array.isArray(comments) && comments.length > 0) {
+        const parts: Doc = [];
+        path.each((comment) => {
+            const commentToken = comment.node;
+            if (commentToken.trailing || commentToken.leading) {
+                return;
+            }
+            const commentDoc = printComment(commentToken, options);
+            if (commentToken.tokenType.name === TokenType.MULTI_LINE_COMMENT) {
+                parts.push(commentDoc);
+            } else {
+                parts.push(lineSuffix(commentDoc));
+            }
+        }, "comments");
+        return [join(hardline, parts)];
+    } else {
+        return [];
+    }
+}
+/**
+ * Prints a comment
+ *
+ * @param comment the comment to print
+ * @param options the formatting options
+ * @returns the formatted comment
+ */
+
+export function printComment(comment: Comment, options: Options): Doc {
+    comment.printed = true;
+    if (comment.tokenType.name === TokenType.MULTI_LINE_COMMENT) {
+        const image = comment.image;
+        const leadingWhitespaceLength = calculateWhitespaceLength(
+            getLeadingWhitespace(options.originalText, comment.startOffset),
+            options.tabWidth
+        );
+        const lines = image
+            .split("\n")
+            .map((line) => decreaseIdentation(line, leadingWhitespaceLength, options.tabWidth));
+        return join(hardline, lines);
+    } else {
+        return comment.image;
+    }
+}
+/**
+ * Gets the leading whitespace of a given offset until the last non-whitespace character or newline
+ *
+ * @param text the text to get the leading whitespace from
+ * @param offset the offset to start from
+ * @returns the leading whitespace
+ */
+function getLeadingWhitespace(text: string, offset: number): string {
+    let start = offset;
+    while (start > 0 && text[start - 1].match(/[^\S\n]/)) {
+        start--;
+    }
+    return text.substring(start, offset);
+}
+/**
+ * Calculates the whitespace length of a given text
+ *
+ * @param text the purely whitespace text
+ * @param tabWidth the width of a tab
+ * @returns the length of the text
+ */
+function calculateWhitespaceLength(text: string, tabWidth: number): number {
+    let length = 0;
+    for (let i = 0; i < text.length; i++) {
+        if (text[i] === "\t") {
+            length += tabWidth;
+        } else {
+            length++;
+        }
+    }
+    return length;
+}
+/**
+ * Decreases the indentation of a given text by a given length
+ * If a tab causes a larger dedentation than desired, spaces are used to replace it
+ *
+ * @param text the text to dedent
+ * @param length the length to dedent
+ * @param tabWidth the width of a tab
+ * @returns the dedented text
+ */
+function decreaseIdentation(text: string, length: number, tabWidth: number): string {
+    let i = 0;
+    while (length > 0 && i < text.length && text[i].match(/[^\S\n]/)) {
+        if (text[i] === "\t") {
+            length -= tabWidth;
```

Comment:
Interesting, the different approaches one takes.
Allowing a mixture of tabs and whitespaces is an interesting choice.
I would have simply only allowed one, and transformed all other whitespace to the chosen one.
Purely hypothetically speaking, of course ;)

Reply:
(Some background info)
I mean the formatter primarily does this.
The one exception are comments: maybe there is a weird reason why you needs tabs in your comment (don't ask me)
The standard prettier behavior is to only really format doc comments, and for all others, you just keep the comment as is.
However, I wanted to at least fix the comments a bit, e.g. when you copy paste code and then have less indentation.
Could I just replace the tabs? Yes
However than I would have needed to write the replacement in a way that only replaces it in leading whitespace, and I was just too lazy for that 😄